File: src/main/scala/com/wikibooks/spark/ch2/scala/PassingFunctionSample.scala
package com.wikibooks.spark.ch2.scala

import org.apache.spark.{SparkConf, SparkContext}

class PassingFunctionSample {

  // 예제 2-1
  var count = 1

  def add1(i: Int): Int = {
    count + 1
  }

  def runMapSample(sc: SparkContext) {
    val rdd1 = sc.parallelize(1 to 10)
    // java.io.NotSerializableException !!!!
    val rdd2 = rdd1.map(add)
    println(rdd2.collect())
  }

  // 예제 2-2
  var increment = 1

  def add(i: Int): Int = {
    i + 1
  }

  def runMapSample2(sc: SparkContext) {
    val rdd1 = sc.parallelize(1 to 10)
    val rdd2 = rdd1.map(Operations.add)
    print(rdd2.collect().toList)
  }

  // 예제 2-5
  def runMapSample3(sc: SparkContext) {
    val rdd1 = sc.parallelize(1 to 10)
    val rdd2 = rdd1.map(_ + increment)
    print(rdd2.collect.toList)
  }

  def runMapSample4(sc: SparkContext) {
    val rdd1 = sc.parallelize(1 to 10)
    val localIncrement = increment
    val rdd2 = rdd1.map(_ + localIncrement)
    print(rdd2.collect().toList)
  }
}

object Operations {
  def add(i: Int): Int = {
    i + 1;
  }
}

object PassingFunctionSampleRunner {

  def main(args: Array[String]) {
    val sc = getSparkContext
    val sample = new PassingFunctionSample

    // 실행할 메서드 주석 제거 후 실행

    // sample.runMapSample(sc)
    // sample.runMapSample2(sc)
    // sample.runMapSample3(sc)
    // sample.runMapSample4(sc)

    sc.stop
  }

  def getSparkContext(): SparkContext = {
    val conf = new SparkConf()
    conf.setMaster("local[*]").setAppName("PassingFunctionSample")
    new SparkContext(conf)
  }
}

File: src/main/java/com/wikibooks/spark/ch2/PassingFunctionSample.java
package com.wikibooks.spark.ch2;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;

import java.util.Arrays;

// 예제 2-3
class Add implements Function<Integer, Integer> {
	@Override
	public Integer call(Integer v1) throws Exception {
		return v1 + 1;
	}
}

public class PassingFunctionSample {

	public void runMapSample(JavaSparkContext sc) {

		JavaRDD<Integer> rdd1 = sc.parallelize(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10));

		JavaRDD<Integer> rdd2 = rdd1.map(new Add());

		System.out.println(rdd2.collect());

		sc.stop();
	}

	public static void main(String[] args) throws Exception {

		JavaSparkContext sc = getSparkContext();

		PassingFunctionSample sample = new PassingFunctionSample();

		sample.runMapSample(sc);

		sc.stop();
	}

	public static JavaSparkContext getSparkContext() {
		SparkConf conf = new SparkConf().setAppName("PassingFunctionSample").setMaster("local[*]");
		return new JavaSparkContext(conf);
	}
}
